core/store/src/trie/mem/arena/mod.rs

```diff
@@ -79,106 +128,90 @@ impl ArenaPos {
     }
 
     fn raw_slice_mut(&mut self, pos: ArenaPos, len: usize) -> &mut [u8] {
         &mut self.chunks[pos.chunk()][pos.pos()..pos.pos() + len]
     }
-
-    /// Provides read access to a region of memory in the arena.
-    pub fn slice<'a>(&'a self, pos: ArenaPos, len: usize) -> ArenaSlice<'a> {
-        ArenaSlice { arena: self, pos, len }
-    }
-
-    /// Provides write access to a region of memory in the arena.
-    pub fn slice_mut<'a>(&'a mut self, pos: ArenaPos, len: usize) -> ArenaSliceMut<'a> {
-        ArenaSliceMut { arena: self, pos, len }
-    }
-
-    /// Represents some position in the arena but without a known length.
-    pub fn ptr<'a>(self: &'a Self, pos: ArenaPos) -> ArenaPtr<'a> {
-        ArenaPtr { arena: self, pos }
-    }
-
-    /// Like `ptr` but with write access.
-    pub fn ptr_mut<'a>(self: &'a mut Self, pos: ArenaPos) -> ArenaPtrMut<'a> {
-        ArenaPtrMut { arena: self, pos }
-    }
 }
 
-impl Arena {
+impl STArena {
     /// Creates a new memory region of the given size to store trie nodes.
     /// The `max_size_in_bytes` can be conservatively large as long as it
     /// can fit into virtual memory (which there are terabytes of). The actual
     /// memory usage will only be as much as is needed.
     pub fn new(name: String) -> Self {
-        Self { memory: ArenaMemory::new(), allocator: Allocator::new(name) }
-    }
-
-    /// Allocates a slice of the given size in the arena.
-    pub fn alloc<'a>(&'a mut self, size: usize) -> ArenaSliceMut<'a> {
-        self.allocator.allocate(&mut self.memory, size)
-    }
-
-    /// Deallocates the given slice from the arena; the slice's `pos` and `len`
-    /// must be the same as an allocation that was returned earlier.
-    pub fn dealloc(&mut self, pos: ArenaPos, len: usize) {
-        self.allocator.deallocate(&mut self.memory, pos, len);
+        Self { memory: STArenaMemory::new(), allocator: Allocator::new(name) }
     }
 
     /// Number of active allocations (alloc calls minus dealloc calls).
     #[cfg(test)]
     pub fn num_active_allocs(&self) -> usize {
         self.allocator.num_active_allocs()
     }
+}
 
-    pub fn memory(&self) -> &ArenaMemory {
+impl Arena for STArena {
+    type Memory = STArenaMemory;
+
+    fn memory(&self) -> &STArenaMemory {
         &self.memory
     }
 
-    pub fn memory_mut(&mut self) -> &mut ArenaMemory {
+    fn memory_mut(&mut self) -> &mut STArenaMemory {
         &mut self.memory
     }
+
+    fn alloc(&mut self, size: usize) -> ArenaSliceMut<Self::Memory> {
+        self.allocator.allocate(&mut self.memory, size)
+    }
+}
+
+impl ArenaWithDealloc for STArena {
+    fn dealloc(&mut self, pos: ArenaPos, len: usize) {
+        self.allocator.deallocate(&mut self.memory, pos, len);
+    }
 }
 
 /// Represents some position in the arena but without a known length.
-#[derive(Clone, Copy)]
-pub struct ArenaPtr<'a> {
-    arena: &'a ArenaMemory,
+#[derive_where(Clone, Copy)]
```

Comment:
FYI: this most likely loses the optimization where `Clone::clone` is identical to copy. In that sense implementing the `Clone` trait manually would be more efficient in terms of compile time (though in a very minor way.)

But more than that, `proc_macro`s are pretty expensive a dependency (e.g. they get linked for host arch even for `check` builds.) If we had it somewhere in our dependency tree from another crate, or if we needed the functionality provided by this proc_macro very extensively, then I think this one would be fine. As is, however, I think the tradeoff is definitely in favour of manual implementations of these traits…

Reply:
Hmm, what's the proper way to manually implement Copy that it would preserve the optimization? (If this question makes sense at all; I'm not sure I understand the optimization).